### PR TITLE
Recognize bounds on impls as const bounds

### PR DESCRIPTION
**Pull request by fee1-dead**
r? @oli-obk

### CODE REVIEW

File: compiler/rustc_hir/src/hir.rs [hunk 1]
```diff
@@ -3064,6 +3064,27 @@ impl<'hir> Node<'hir> {
...
+
+    /// Returns `Constness::Const` when this node is a const fn/impl.
+    pub fn constness(&self) -> Constness {
+        match self {
+            Node::Item(Item {
+                kind: ItemKind::Fn(FnSig { header: FnHeader { constness, .. }, .. }, ..),
+                ..
+            })
+            | Node::TraitItem(TraitItem {
+                kind: TraitItemKind::Fn(FnSig { header: FnHeader { constness, .. }, .. }, ..),
```
Comment (oli-obk) [inline compiler/rustc_hir/src/hir.rs:3076]: this cannot happen (yet), maybe put it in a separate match arm and assert that the constness is `NotConst`

**Comment by oli-obk** [comment]
Nice! This is a great fix.

File: compiler/rustc_hir/src/hir.rs [hunk 2]
```diff
@@ -3064,6 +3064,35 @@ impl<'hir> Node<'hir> {
...
+                ..
+            })
+            | Node::Item(Item { kind: ItemKind::Impl(Impl { constness, .. }), .. }) => *constness,
+
+            Node::TraitItem(TraitItem {
+                kind: TraitItemKind::Fn(FnSig { header: FnHeader { constness, .. }, .. }, ..),
+                ..
+            })
+            | Node::ImplItem(ImplItem {
+                kind: ImplItemKind::Fn(FnSig { header: FnHeader { constness, .. }, .. }, ..),
```
Comment (oli-obk) [inline compiler/rustc_hir/src/hir.rs:3082]: this on the other hand can happen for inherent impls, as we already have stable inherent const fns

Comment (fee1-dead) [reply compiler/rustc_hir/src/hir.rs:3082]: Okay, fixing that now.

**Comment by oli-obk** [comment]
looks like I was wrong, it is reachable, but only in code that already errors. I guess go back to your original PR. Sorry about that.

File: compiler/rustc_hir/src/hir.rs (hunk 1 above)
Comment (fee1-dead) [reply compiler/rustc_hir/src/hir.rs:3082]: I'll just revert this change since it is caught elsewhere.

**Pull request merged**